flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/MiniDispatcher.java

```diff
@@ -124,16 +125,23 @@ protected CleanupJobState jobReachedTerminalState(ExecutionGraphInfo executionGr
                 executionGraphInfo.getArchivedExecutionGraph();
         final CleanupJobState cleanupHAState = super.jobReachedTerminalState(executionGraphInfo);
 
-        if (jobCancelled || executionMode == ClusterEntrypoint.ExecutionMode.DETACHED) {
+        JobStatus jobStatus = archivedExecutionGraph.getState();
+        if ((jobStatus != null && jobStatus.isGloballyTerminalState())
+                && (jobCancelled || executionMode == ClusterEntrypoint.ExecutionMode.DETACHED)) {
             // shut down if job is cancelled or we don't have to wait for the execution result
             // retrieval
             log.info(
                     "Shutting down cluster with state {}, jobCancelled: {}, executionMode: {}",
-                    archivedExecutionGraph.getState(),
+                    jobStatus,
+                    jobCancelled,
+                    executionMode);
+            shutDownFuture.complete(ApplicationStatus.fromJobStatus(jobStatus));
+        } else {
+            log.warn(
```

Comment:
Why do we need to print a warning here?

Always try to think about it from the user perspective. If users sees this log message, is it relevant to what's going on? Is it something he / she should investigate?